File: client/python/inputsystemwarriors-test.py
from gamium import *
import sys


gamium = GamiumClient("127.0.0.1", 50061)
gamium.connect()
ui = gamium.ui()


def mouse_keyboard_test():
    screen = gamium.screen()
    for _ in range(3):
        gamium.actions().click(Vector2(screen.width / 2, screen.height / 2)).perform()
        gamium.send_key(KeyBy.unity_keycode("W"), SendKeyOptions(300))
        gamium.send_key(KeyBy.unity_keycode("A"), SendKeyOptions(300))
        gamium.send_key(KeyBy.unity_keycode("S"), SendKeyOptions(300))
        gamium.send_key(KeyBy.unity_keycode("D"), SendKeyOptions(300))


mouse_keyboard_test()


def settings_test():
    gamium.send_key(KeyBy.unity_keycode("P"), SendKeyOptions(300))

    buttonPaths = [
        "/UI[1]/UI - Menu[1]/UI Menu - Canvas[1]/Panel_MiddleStrip[1]/Panel_Left[1]/VerticalLayout_ButtonOptions[1]/UI_Button_Option_Rebind[1]",
        "/UI[1]/UI - Menu[1]/UI Menu - Canvas[1]/Panel_MiddleStrip[1]/Panel_Left[1]/VerticalLayout_ButtonOptions[1]/UI_Button_Option_Quit[1]",
        "/UI[1]/UI - Menu[1]/UI Menu - Canvas[1]/Panel_MiddleStrip[1]/Panel_Left[1]/VerticalLayout_ButtonOptions[1]/UI_Button_Option_About[1]",
    ]

    for path in buttonPaths:
        ui.click(By.path(path))

    gamium.send_key(KeyBy.unity_keycode("P"), SendKeyOptions(300))


settings_test()


def keyboard_test():
    keys = []
    for attr_name in dir(UnityKeyboard):
        if not attr_name.startswith("__"):
            keys.append(attr_name)

    for key in keys:
        gamium.send_key(KeyBy.unity_keyboard(key))
        text = ui.get_text(By.path("/UI[1]/Canvas[1]/Text (TMP)[1]"))
        assert getattr(UnityKeyboard, text) == getattr(
            UnityKeyboard, key
        ), f"keyboard test failed: {getattr(UnityKeyCode, text)} != {getattr(UnityKeyCode, key)}"


keyboard_test()


def quit():
    gamium.actions().app_quit().perform()
    sys.exit(0)


quit()

File: client/typescript/test/inputsystemwarriors-test.ts
import assert from 'assert';
import { By, GamiumClient, KeyBy, NodeGamiumService, Player, UnityKeyboard } from 'gamium';
import { test } from './functions';

(async () => {
  const gamiumService = new NodeGamiumService('127.0.0.1', 50061);
  await gamiumService.connect();
  const gamium = new GamiumClient(gamiumService);
  const ui = gamium.ui();

  await test('Mouse Keyboard test', async () => {
    const screen = await gamium.screen();
    for (let index = 0; index < 3; index++) {
      await gamium
        .actions()
        .click({ x: screen.width / 2, y: screen.height / 2 })
        .perform();
      await gamium.sendKey(KeyBy.unityKeycode('W'), { duratiomMs: 300 });
      await gamium.sendKey(KeyBy.unityKeycode('A'), { duratiomMs: 300 });
      await gamium.sendKey(KeyBy.unityKeycode('S'), { duratiomMs: 300 });
      await gamium.sendKey(KeyBy.unityKeycode('D'), { duratiomMs: 300 });
    }
  });

  await test('Settings test', async () => {
    await gamium.sendKey(KeyBy.unityKeycode('P'), { duratiomMs: 300 });

    const buttonPaths = [
      '/UI[1]/UI - Menu[1]/UI Menu - Canvas[1]/Panel_MiddleStrip[1]/Panel_Left[1]/VerticalLayout_ButtonOptions[1]/UI_Button_Option_Rebind[1]',
      '/UI[1]/UI - Menu[1]/UI Menu - Canvas[1]/Panel_MiddleStrip[1]/Panel_Left[1]/VerticalLayout_ButtonOptions[1]/UI_Button_Option_Quit[1]',
      '/UI[1]/UI - Menu[1]/UI Menu - Canvas[1]/Panel_MiddleStrip[1]/Panel_Left[1]/VerticalLayout_ButtonOptions[1]/UI_Button_Option_About[1]',
    ];

    for (const path of buttonPaths) {
      await ui.click(By.path(path));
    }

    await gamium.sendKey(KeyBy.unityKeycode('P'), { duratiomMs: 300 });
  });

  await test('Keyboard test', async () => {
    const keys = Object.keys(UnityKeyboard).filter((v) => isNaN(Number(v)));

    for (const key of keys) {
      await gamium.sendKey(KeyBy.unityKeyboard(key as keyof typeof UnityKeyboard));
      const text = await ui.getText(By.path('/UI[1]/Canvas[1]/Text (TMP)[1]'));
      assert.deepStrictEqual(UnityKeyboard[text], UnityKeyboard[key], `fail ${UnityKeyboard[key]} is not equal to ${UnityKeyboard[text]}`);
    }
  });

  await test('Quit', async () => {
    await gamium.actions().appQuit().perform();
    process.exit(0);
  });
})().catch((e) => {
  console.error(e);
  process.exit(1);
});
